### DISCUSSION TITLE
Is there a way to add images to corner controls with Fabric JS?

### DISCUSSION THREAD

**Discussion by ibrahimrehman1** [fabricjs/fabric.js#8397] [Q&A] [answered]
Hi Guys!

I want to add images to the corners like in the below image when using fabric.IText() or when adding any object to canvas. I haven't been able to find a solution to this. Kindly help me with this. Thanks!

I am using the following CDN:

```
<script
      src="https://cdnjs.cloudflare.com/ajax/libs/fabric.js/3.6.3/fabric.min.js"
      crossorigin="anonymous"
    ></script>

```

This is my code so far:

```

new fabric.IText("Default Text", {
        
        fontFamily: "sans-serif",
        fontSize: 90,
        backgroundColor: "red",
        rotatingPointOffset: 120,
        cornerColor: "blue",
        cornerSize: 50,
        left: 250,
        top: 250,
        angle: 0,
        padding: 40,
        width: width,
        height: height,
        fill: "white",
        originX: "center",
        originY: "center",
      });

```

![WhatsApp Image 2022-10-27 at 9 37 34 PM (1)](https://user-images.githubusercontent.com/63722045/198507206-58b45a09-7588-4a37-acff-5059a18c98b2.jpeg)

**Comment by ShaMan123** [comment, marked as answer]
Yes
For sure
Look at the control logic.
There is a render method.
http://fabricjs.com/custom-control-render

**Reply by ibrahimrehman1** [reply]
@ShaMan123 Thanks! it looks like it will work, but is there a solution in fabric.js 3.6.3 to achieve the same thing? I am actually using this version in my application

**Reply by ShaMan123** [reply]
That is a bad idea altogether IMO
V4 is considered not supported
V3 is dead

**Reply by ibrahimrehman1** [reply]
Ok, I understand. I should migrate then. Thanks, @ShaMan123 for your help 👍

**Reply by shashank-brightness** [reply]
@ShaMan123 Is there any way we can do this only for specific object and not for others? fabric.Object.prototype.controls seems to override it for all.

Also, isn't there any way just to replace image for the default controls i.e. tl, tr, bl, br etc....? because even if i try to create let's say new control with same function i am not getting which actionHandler i should provide for my new control with different image

I have this code for adding cropTopLeft control which could indicate cropping

```
addCropRectangleControls(cropSelectionRect) {
    let self = this;
    return new Promise((resolve,reject) => {
        fabric.Image.fromURL(getObjectControlImages().cropTopLeftImage,function(img, isError) {
            if(isError) {
                reject(isError);
            }
            img.scaleToWidth(20);
            const cropTopLeftIconDataUrl = img.toDataURL();
            const cropTopLeftIcon = document.createElement('img');
            
            cropTopLeftIcon.src = cropTopLeftIconDataUrl;

            console.log(cropTopLeftIconDataUrl)

            const cropTopLeftControl = new fabric.Control({
                x: -0.5,
                y: -0.5,
                offsetY: 8,
                offsetX: 8,
                //actionHandler: fabric.controlsUtils.scaleCursorStyleHandle, 
                cursorStyleHandler: fabric.controlsUtils.scaleCursorStyleHandler,
                actionName: 'scale',
                render: self.renderIcon(cropTopLeftIcon),
                cornerSize: 20
            });
            cropSelectionRect.setControlsVisibility({
                cropTopLeft: true,
                tl: false
            })
            fabric.Object.prototype.controls.cropTopLeftControl = cropTopLeftControl;
            resolve();
        }, {crossOrigin: 'anonymous'});
    })
}
```
and result is 
![Screenshot 2024-09-05 113525](https://github.com/user-attachments/assets/422611e6-2494-48e5-a2be-ce049b422ca9)


like this but nothing happens when i click on newly added top left control. `actionHandler` should be what so that it works same as the default `tl` control?

**Reply by asturur** [reply]
The default top left control has the scale action.
You find it here:
```
import { controlsUtils } from 'fabric';

controlsUtils.scalingEqually

```

**Reply by shashank-brightness** [reply]
@asturur Thanks for the answer. Actually i did the same earlier and found it working but now there is one more issue that when i drag left top corner, it starts scaling like it does from right bottom. i.e. holding and then going towards more top left make shape even smaller and not larger. 
This thing even gets fixed if i change `originX` and `originY` property to `bottom` and `right` but the thing is that i never changed those two props which defaults them to `left` and `top`. All four corners uses the same `scalingEqually` method.
What could be the issue?

**Reply by asturur** [reply]
The scaling equally is wrapped with the anchor position wrapper.
How does your control looks like? can you paste the whole action?

**Reply by shashank-brightness** [reply]
Yes i have assigned `actionHandler` to direct `fabric.controlsUtils.scalingEqually` as i am still using 5th version via cdn so not used import for controlsUtils.

**Reply by asturur** [reply]
it should be good enough, providing that your transform is actually doing what is supposed to do.
Does your control have an actionName too? which is it? can you console.log during the 'scaling' event what the options passed to your event handler look like?

**Reply by shashank-brightness** [reply]
I don't have `actionName` for my control and this is how options look when i attach scaling handler for the object where i set that newly made control visible and used.
![Screenshot 2024-09-12 174756](https://github.com/user-attachments/assets/be2de4d6-351e-4e60-abde-841b140dad65)
![Screenshot 2024-09-12 172135](https://github.com/user-attachments/assets/160344ee-5464-4bef-a550-9c0905398ec2)

and this is the whole code for control. I have no idea why it reverses the scale operation.
```
addCropRectangleControls(cropSelectionRect) {
        let self = this;
        return new Promise((resolve,reject) => {
            fabric.Image.fromURL(getObjectControlImages().cropTopLeftImage,function(img, isError) {
                if(isError) {
                    reject(isError);
                }
                img.scaleToWidth(20);
                const cropTopLeftIconDataUrl = img.toDataURL();
                const cropTopLeftIcon = document.createElement('img');
                
                cropTopLeftIcon.src = cropTopLeftIconDataUrl;
    
                const cropTopLeftControl = new fabric.Control({
                    x: -0.5,
                    y: -0.5,
                    cornerSize: 20,
                    offsetY: 8,
                    offsetX: 8,
                    actionHandler: fabric.controlsUtils.scalingEqually, 
                    cursorStyleHandler: fabric.Object.prototype.controls.tl.cursorStyleHandler,
                    render: self.renderIcon(cropTopLeftIcon),
                    
                });
                cropSelectionRect.setControlsVisibility({
                    cropTopLeft: true,
                    tl: false
                })
                fabric.Object.prototype.controls.cropTopLeftControl = cropTopLeftControl;
                resolve();
            }, {crossOrigin: 'anonymous'});
        })
    },
```

**Reply by shashank-brightness** [reply]
@asturur Can you please suggest on above? How do I need to wrap `scalingEqually` method to scale in same way as older controls on each newly added controls?
